I13 ekf slam

nanwei1 opened the pull request:
Files are updated in i13/code v1.0/ME597-6-EKFSLAM

Changes made:
- Functionalized measurement and motion models
- Linearized models were also functionalized
- The main difference from the previous version (ekf_slam) is that the map size is updated as new features are detected:
  - Initially the states and covariance only contain information about the vehicle (x,y,theta in this case)
  - As new features are observed, the estimate of this feature's state is augmented right below the vehicle states. The feature list, measurements, etc are rearranged accordingly
  - The covariance for the newly observed feature is set to a predefined value (S0mm=eye(2)), and the off-diagonal terms are set to 0
  - A line of text is printed to the Command Window, indicating how many features have been detected
-  Added examples of vehicle running with different trajectories in different maps (ekf_slam_lineMap and ekf_slam_circularMap)
- The EKF function will be incorporated in the next round
- Note that just like the original code, this simulation fails from time to time as the linearization may not be accurate or the covariance matrix becomes not positive-definite. This code warns the user when the update in vehicle position is too large compared to input velocity but cannot deal with these failures. Any comments on how to fix this will be appreciated.
- Added example where features are placed further way from vehicle trajectory (ekf_slam_largerCircularMap) to show that linearization works better in this case (as illustrated by Mapping II slide 28). If you try these simulation enough times, you will see that this one fails less frequently than ekf_slam_circularMap. Below is a comparison of the estimates:
  ![radius 6](https://cloud.githubusercontent.com/assets/17601506/13560533/4f5efb6a-e3ee-11e5-9133-e77b5b467484.jpg)
  (Features in a r=6m circle)

![radius 12](https://cloud.githubusercontent.com/assets/17601506/13560540/571330b0-e3ee-11e5-9d08-20d39a94427a.jpg)
(Features in a r=12m circle)

nanwei1 committed:
Updated EKA_SLAM (3.6,2016)

- Functionalized measurement and motion models
- Linearized models were also functionalized
- The main difference from the previous version is that the map size is updated was new features are detected:
* Initially the states and covariance only contain information for the vehicle (x,y,theta in this case)
* As new features are observed, the estimate of this feature's state is augmented right below the vehicle states. The feature list, measurements, etc are rearranged accordingly
* The covariance for the newly observed feature is set to a predefined value (S0mm=eye(2)), and the off-diagonal terms are set to 0
* A line of text is printed to the Command Window, indicating how many features have been detected

- Note that just like the original code, this simulation fails from time to time as the linearization may not be accurate or the covariance matrix becomes not positive-definite. This code warns the user when the update in vehicle position is too large compared to input velocity but cannot deal with these failures. Any comments on how to fix this will be appreciated.

nanwei1 committed:
Create two_wheel_motion_model.m

nanwei1 committed:
Create two_wheel_motion_linearized_model.m

nanwei1 committed:
Create range_bearing_meas_model.m

nanwei1 committed:
Create range_bearing_meas_linearized_model.m

nanwei1 committed:
Create rearrangeMap.m

nanwei1 committed:
ekf_slam_circularMap

Simulation with a map with 6 features in a circle with 6m radius.

nanwei1 committed:
Create ekf_slam_lineMap

Simulation of EKF SLAM with a map that has features along 2 straight lines. The vehicle moves back and forth in this map.

nanwei1 committed:
Rename ekf_slam_lineMap to ekf_slam_lineMap.m

nanwei1 committed:
Update ekf_slam_lineMap.m

nanwei1 committed:
Create ekf_slam_largerCircularMap.m

This simulation if the same as ekf_slam_circularMap except for:
- The features are further away from the vehicle trajectory
- Max range of the sensor is extended so it works with this map

The purpose of this simulation is to show that the estimation of vehicle state is better compared to ekf_slam_circularMap because the features are further away from the vehicle, and thus linearization is a better approximation in this case (as illustrated by Mapping II slide 28)

rickeywang committed:
Code Review - Very Good

The new examples work really well. It let me saw the difference that
feature distances made.

Tidy up code, and spelling. Mostly performance/cosmetic changes.

The three examples have a lot of duplicate code for running the simulation. Perhaps they can be
merged into one? Not sure how this would work though.

cbchau commented:
Suggest putting the plotting code in separate functions which can be used for multiple scripts.
![plot code](https://cloud.githubusercontent.com/assets/16365243/13892500/378d9c78-ed2e-11e5-9998-de87d01396bd.PNG)

nanwei1 committed:
Update eka_slam

- Combined all 4 examples into 1 file
- An ekfSLAMplot function is used to replace the original ~40 lines for plotting
- In case if S (the covariance matrix) gets not positive definite, manually add a diagonal matrix to make it positive definite to keep the code running
- I considered using the ekf function, but the issue was that this ekf_slam updates the size of its map and covariance matrix as new features are discovered. Initialization of newly discovered features is performed after the prediction step using the predicted robot states, and the size of mu and S is updated before the measurement update. Thus, I did not find a way to directly use the ekf function without much modification

nanwei1 committed:
Delete ekf_slam_circularMap.m

nanwei1 committed:
Delete ekf_slam_largerCircularMap.m

nanwei1 committed:
Delete ekf_slam_lineMap.m

nanwei1 committed:
Create ekfSLAMplot.m

nanwei1 commented:
- Combined all 4 examples into 1 file. The user will need to pick an example at the beginning of the code.
- An ekfSLAMplot function is used to replace the original ~40 lines for plotting
- In case if S (the covariance matrix) gets not positive definite, manually add a diagonal matrix to make it positive definite to keep the code running. Warn the use when this happens
- I considered using the ekf function, but the issue was that this ekf_slam code updates the size of its map and covariance matrix as new features are discovered. Initialization of newly discovered features is performed after the prediction step using the predicted robot states, and the size of mu and S is updated before the measurement update. Thus, I did not find a way to directly use the ekf function without much modification
- Range/bearing only measurement model was tested. I did not upload the code because it basically did not work. The problem began with initialization of features; with only 1 measurement, we needed to estimate the (x,y) position of a feature, which results in poor initialization that messed up the whole simulation

chutsu commented on matlab_simulation/code v1.0/ME597-6-EKFSLAM/ekf_slam.m:
```diff
@@ -113,102 +116,94 @@
-            S = (eye(n+2*M)-K*Ht)*S;
+            S = (eye(length(S))-K*Ht)*S;
+            
+            % In cases if S bemoes not positive definite, manually make it
```
minor edit, suggest 'is not positive definite' over 'becomes not positive definite'

chutsu commented on matlab_simulation/code v1.0/ME597-6-EKFSLAM/range_bearing_meas_linearized_model.m:
```diff
@@ -0,0 +1,25 @@
+            N=length(mu);
+            Fi = zeros(5,N);
+            Fi(1:3,1:3) = eye(3);
+            Fi(4:5,3+2*(i-1)+1:3+2*i) = eye(2);
```
some explaination of how this `Fi` is used would be nice.

chutsu replied:
like so:

```
% feature selector vector
Fi = zeros(5, N);   % N is number of elements (num of vehicle states + 2 * num features)
Fi(1:n, 1:n) = eye(n);  % robot x, y, z
Fi(4:5, 3+2*(i-1)+1:3+2*i) = eye(2);  % measurement y1, y2
```

stevenwaslander replied:
```
Fi(1:n, 1:n) = eye(n);  % robot x, y, theta
Fi(4:5, 3+2*(i-1)+1:3+2*i) = eye(2);  % feature x, y
```

chutsu commented on matlab_simulation/code v1.0/ME597-6-EKFSLAM/range_bearing_meas_linearized_model.m:
```diff
@@ -0,0 +1,25 @@
+                dx/rp^2]*Fi;
+            
+             h_mu_bar = [rp;
+                (atan2(dy,dx) - mu(3))];
```
`h_mu_bar` is not used anywhere.

chutsu commented on matlab_simulation/code v1.0/ME597-6-EKFSLAM/two_wheel_motion_linearized_model.m:
```diff
@@ -0,0 +1,10 @@
+function Gt = two_wheel_motion_linearized_model(x,u,dt)
+% linearized motion model of a 2-wheel robot
+% can be found P.20 of Mapping II slides
```
Avoid referencing slides, they can change
